libraries/chain/block_header_state_legacy.cpp

```diff
@@ -209,12 +216,15 @@ namespace eosio::chain {
       }
 
       if (new_finalizer_policy) {
-         new_finalizer_policy->generation = 1;
+         new_finalizer_policy->generation = 1; // only allowed to be set once
          // set current block_num as qc_claim.last_qc_block_num in the IF extension
          qc_claim_t initial_if_claim { .block_num = block_num,
                                        .is_strong_qc = false };
          emplace_extension(h.header_extensions, instant_finality_extension::extension_id(),
                            fc::raw::pack(instant_finality_extension{ initial_if_claim, std::move(new_finalizer_policy), {} }));
+      } else if (qc_claim) {
+         emplace_extension(h.header_extensions, instant_finality_extension::extension_id(),
+                           fc::raw::pack(instant_finality_extension{ *qc_claim, {}, {} }));
```

Comment:
It looks like we keep propagaing the first qc_claim we got? Why is this needed at all?

Reply:
Because every block after IF Genesis Block is suppose to have an IF extension.